### PR TITLE
Update JSDoc and align custom JSDoc plugins

### PR DESCRIPTION
**Pull request by ma2ciek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature (jsdoc-plugins): Updated JSDoc and align custom JSDoc plugins. Part of ckeditor/ckeditor5#7575.

* Improved performance,
* Fixed incorrectly and inconsistently documented functions. Previously they were marked as static/instance using the `.` or `#` symbols. Now they are documented the same as classes, interfaces, and mixins using the `~` symbol. (e.g. `module:widget/utils~toWidget`),
* Validator was improved, now it detects duplicated modules and previously hidden mistakes.
* Updated JSDoc allows for modern JS syntax, `async/await` among others.

MAJOR BREAKING CHANGE (jsdoc-plugins): Functions documented in a module will no longer be documented as static functions. Therefore links to functions will not contain the `static-` part and links created in the past will no longer work.

---

Requires:

*   ckeditor/ckeditor5#7642
*   cksource/umberto#870

### CODE REVIEW

**Comment by ma2ciek** [comment]
@pomek Could you make a review?

File: packages/jsdoc-plugins/lib/relation-fixer/filteroutintenraldoclets.js [hunk 1]
```diff
@@ -11,25 +11,11 @@
...
-	// Filter out exported symbols that are defined previously (JSDoc creates doclets for both).
-	const isUnwanted = ( doclet.scope == 'inner' && doclet.kind == 'constant' && doclet.undocumented );
+		.filter( doclet => doclet.memberof != '<anonymous>' );
 
-	if ( isUnwanted ) {
-		if ( doclets.find( d => doclet.longname === d.longname && d !== doclet ) ) {
-			return false;
-		}
-	}
+	// return filterOutReExportedSymbols( doclets );
```
Comment (pomek) [inline packages/jsdoc-plugins/lib/relation-fixer/filteroutintenraldoclets.js:18]: What about that?

File: packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js [hunk 2]
```diff
@@ -75,41 +75,46 @@ function getMissingDocletsData( docletCollection, interfaceClassOrMixinDoclet, o
  * @param {DocletCollection} docletCollection
  * @param {Doclet} childDoclet
  * @param {Options} options
+ * @returns {Doclet[]}
```
Comment (pomek) [inline packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js:78]:
```suggestion
 * @returns {Array.<Doclet>}
```

And below.

File: packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js [hunk 3]
```diff
@@ -75,41 +75,46 @@ function getMissingDocletsData( docletCollection, interfaceClassOrMixinDoclet, o
...
-			}
-		}
+	/** @type {Doclet[]} */
+	const docletsToAdd = [];
 
-		return true;
+	for ( const ancestor of ancestors ) {
+		docletsToAdd.push(
+			...docletCollection.get( `memberof:${ ancestor }` )
+				.filter( doclet => shouldDocletBeAdded( doclet, options ) )
```
Comment (pomek) [inline packages/jsdoc-plugins/lib/relation-fixer/getmissingdocletsdata.js:90]:
```suggestion
			...docletCollection.get( `memberof:${ ancestor }` ).filter( doclet => shouldDocletBeAdded( doclet, options ) )
```

or

```suggestion
			...docletCollection
				.get( `memberof:${ ancestor }` )
				.filter( doclet => shouldDocletBeAdded( doclet, options ) )
```

File: packages/jsdoc-plugins/lib/relation-fixer/addmissingdoclets.js [hunk 4]
```diff
@@ -90,8 +86,59 @@ function addMissingDoclets( doclets ) {
...
+			// If the doclet has inherited property don't output it
+			// as it should be replaced by the parent's class/interface/mixin method doclet.
+			if ( willDocletBeAdded && doclet.inheritdoc === undefined ) {
+				return false;
+			}
+
+			return true;
+		} ),
+		// Do not output doclets for doclets having its own documentation.
+		...[ ...docletToAddMap.values() ].filter( doclet => {
```
[the remark this answers is not in the document]
Comment (ma2ciek) [reply packages/jsdoc-plugins/lib/relation-fixer/addmissingdoclets.js:112]: I'll rephrase it.

**Comment by ma2ciek** [comment]
I've fixed the mentioned issues and merged master onto the PR branch. Could you do the final check @pomek?

**Pull request merged**